core/src/test/scala/unit/kafka/log/LogCleanerTest.scala

```diff
@@ -1213,6 +1213,51 @@ class LogCleanerTest extends JUnitSuite {
     assertEquals("The tombstone should be retained.", 1, log.logSegments.head.log.batches.iterator.next().lastOffset)
   }
 
+  /**
+    * Verify that cleaner does not crash, and handles the case where a log segment has messages
+    * with offsets that overflow base_offset + Integer.MAX_VALUE due to KAFKA-5413.
+    */
+  @Test
+  def testLogSegmentWithIndexOffsetOverflow(): Unit = {
+    val tp = new TopicPartition("test", 0)
+    val cleaner = makeCleaner(Int.MaxValue)
+    val logProps = new Properties()
+    logProps.put(LogConfig.SegmentBytesProp, Integer.MAX_VALUE: java.lang.Integer)
+    val log = makeLog(config = LogConfig.fromProps(logConfig.originals, logProps))
+
+    val keys = List(1, 2, 1, 2)
+    val values = List(1, 2, 1, 2)
+    val offsets = List(0L, 1L, Integer.MAX_VALUE + 2L, Integer.MAX_VALUE + 3L)
+
+    // Create a "legacy" log segment, consisting of messages that would result in index overflow
+    var i = 0
+    while (i < offsets.length) {
+      val set = MemoryRecords.withRecords(offsets(i), CompressionType.NONE, 0,
+                                          new SimpleRecord(keys(i).toString.getBytes,
+                                                           values(i).toString.getBytes))
+      if (i == 0)
+        log.appendAsFollower(set)        // write the first record using log.append API
+      else
+        log.activeSegment.log.append(set)  // write all records directly to the file
```

Comment:
I'm wondering if we could construct the FileRecords prior to building the Log. The problem is that writing to the file directly will result in inconsistent log state. But if we do it up-front before constructing the Log, then it should be ok.

Reply:
I guess we may as well use the new api in `LogSegment` which allows us to write and skip the index append.